### PR TITLE
add expand/collapse buttons to search results

### LINKED ISSUE #1185: Expand/compact view

### PR DESCRIPTION
**Pull request by dnoneill**
Screenshot 2024-08-16 at 2 43 30 PM

closes #1185

### CODE REVIEW

File: app/assets/stylesheets/earthworks.css [hunk 1]
```diff
@@ -152,3 +152,10 @@ Uncomment lines 142-144
     display: none;
   }
 }
+
+.expand-collapse .btn:disabled {
```
Comment (jcoyne) [inline app/assets/stylesheets/earthworks.css:156]: I don't think you need the `:disabled` pseudo selector here.

Comment (dnoneill) [reply app/assets/stylesheets/earthworks.css:156]: I am not sure I understand your meaning. I want these styles to show up only when the button is disabled.

Comment (jcoyne) [reply app/assets/stylesheets/earthworks.css:156]: They will. You set variables for disabled styles.

File: app/views/catalog/_sort_and_per_page.html.erb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<div id="sortAndPerPage" class="sort-pagination d-md-flex justify-content-between" role="navigation" aria-label="<%= t('blacklight.search.per_page.aria_label')%>">
+  <%= render partial: "paginate_compact", object: @response if show_pagination? %>
```
Comment (jcoyne) [inline app/views/catalog/_sort_and_per_page.html.erb:2]: if you use "partial", don't you also need to use "locals". I prefer to omit partial.

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/javascript/controllers/expandcollapse_controller.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+export default class extends Controller {
+  toggleDescriptions(evt) {
+    const activeClass = 'active'
+    const currentActive = this.element.querySelector(`.${activeClass}`)
+    currentActive.classList.remove(activeClass)
+    currentActive.disabled = false;
+    const target = evt.target.classList.contains('bi') ? evt.target.parentElement : evt.target;
+    target.disabled = true;
+    target.classList.add(activeClass)
+    const elements = Array.from(document.getElementsByClassName('more-info-area'));
```
Comment (jcoyne) [inline app/javascript/controllers/expandcollapse_controller.js:12]: This is a pretty fragile way to bind to these elements.  Have you considered making them stimulus targets for this controller?  Alternatively, you could use stimulus outlets to reference another controller responsible for each document. https://stimulus.hotwired.dev/reference/outlets or even just emit an custom event like "showList"/"showParagraph", that another controller will bind to.

File: app/views/catalog/_sort_and_per_page.html.erb [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  <div class="search-widgets">
+    <%= render_results_collection_tools %>
+    <div class="expand-collapse btn-group" data-controller="expandcollapse">
+      <button class="btn btn-outline-secondary active" disabled data-action="expandcollapse#toggleDescriptions"><i class="bi bi-list-ul"></i></button>
```
Comment (jcoyne) [inline app/views/catalog/_sort_and_per_page.html.erb:6]: These buttons don't look accessible to me.

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/views/catalog/_sort_and_per_page.html.erb (hunk 2 above)
Comment (dnoneill) [reply app/views/catalog/_sort_and_per_page.html.erb:2]: This is copied directly from blacklight.

File: app/assets/stylesheets/earthworks.css (hunk 1 above)
Comment (dnoneill) [reply app/assets/stylesheets/earthworks.css:156]: Gotcha. That makes sense.

File: app/javascript/controllers/expandcollapse_controller.js (hunk 3 above)
Comment (dnoneill) [reply app/javascript/controllers/expandcollapse_controller.js:12]: I thought about the stimulus targets but then you have to figure out which got clicked. I will think about it some more maybe.

File: app/views/catalog/_sort_and_per_page.html.erb (hunk 2 above)
Comment (jcoyne) [reply app/views/catalog/_sort_and_per_page.html.erb:2]: I was wrong. This is actually important.

File: app/javascript/controllers/expandcollapse_controller.js (hunk 3 above)
Comment (jcoyne) [reply app/javascript/controllers/expandcollapse_controller.js:11]: I still think that binding behaviors to css styles (`.more-info-area`) is not very robust.  I urge you to look at the solutions that are built into Stimulus for this.

File: app/views/catalog/_sort_and_per_page.html.erb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<div id="sortAndPerPage" class="sort-pagination d-md-flex justify-content-between" role="navigation" aria-label="<%= t('blacklight.search.per_page.aria_label')%>">
+  <%= render partial: "paginate_compact", object: @response if show_pagination? %>
+  <div class="search-widgets">
+    <%= render_results_collection_tools %>
+    <div class="expand-collapse btn-group" data-controller="expandcollapse">
+      <button class="btn btn-outline-secondary active" aria-label="collapse descriptions" disabled data-action="expandcollapse#toggleDescriptions">
```
Comment (jcoyne) [inline app/views/catalog/_sort_and_per_page.html.erb:6]: Shouldn't these also have `aria-controls`?  If one of them always has to be selected, can this just be a styled radio button (e.g. hide the button entirely and display a image in the label)?

File: app/javascript/controllers/expandcollapse_controller.js (hunk 3 above)
Comment (dnoneill) [reply app/javascript/controllers/expandcollapse_controller.js:11]: Gotcha I misunderstood the area the comment was pointing towards, for some reason my brain said the top line and not the last line. Makes sense.

File: app/views/catalog/_sort_and_per_page.html.erb (hunk 5 above)
Comment (dnoneill) [reply app/views/catalog/_sort_and_per_page.html.erb:6]: Please see https://github.com/sul-dlss/earthworks/pull/1234/files to see if this is what you meant for both this and the previous comment. Also if you know the css off the top of your head for styling this I would appreciate it. What is in there now doesn't work.

**Pull request closed without merging**